plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucketServerPullRequest.ts

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { InputError } from '@backstage/errors';
+import {
+  getBitbucketServerRequestOptions,
+  ScmIntegrationRegistry,
+} from '@backstage/integration';
+import { createTemplateAction } from '@backstage/plugin-scaffolder-node';
+import fetch, { RequestInit, Response } from 'node-fetch';
+import { parseRepoUrl } from './util';
+import { Config } from '@backstage/config';
+
+const createPullRequest = async (opts: {
+  project: string;
+  repo: string;
+  title: string;
+  description?: string;
+  toRef: object;
+  fromRef: object;
```

Comment:
maybe a bit unorthodox to have these as objects, rather than a precise type

Reply:
changed it.